File: tape/sdk/go/chaos/snapshot.go
package chaos

import (
	"context"
	"encoding/json"
	"sort"
	"strings"
	"time"

	tape "github.com/vamsiramakrishnan/durable-agents/tape/sdk/go"
	"github.com/vamsiramakrishnan/durable-agents/tape/sdk/go/tapepb"
)

// Keys stripped before comparing canonical journals. Kept in lockstep
// with `tape.chaos.snapshot._STRIP_KEYS` (Python).
var stripKeys = map[string]struct{}{
	"ts_ms": {}, "started_at_ms": {}, "ended_at_ms": {}, "last_update_time_ms": {},
	"lease_expires_at_ms": {}, "claim_expires_at_ms": {}, "dispatch_claim_expires_at_ms": {},
	"next_dispatch_at_ms": {}, "next_attempt_at_ms": {}, "fire_at_ms": {},
	"lease_owner": {}, "claimed_by": {}, "dispatch_claimed_by": {},
	"trace_id": {}, "span_id": {}, "parent_span_id": {},
	"seq": {}, "global_seq": {},
	"invocation_id": {},
}

var terminalRunStatuses = map[string]struct{}{
	"terminal": {}, "failed": {}, "cancelled": {}, "stuck": {},
}

func canonical(value any, runIDMap map[string]string) any {
	switch v := value.(type) {
	case map[string]any:
		out := map[string]any{}
		for k, vv := range v {
			if _, drop := stripKeys[k]; drop {
				continue
			}
			out[k] = canonical(vv, runIDMap)
		}
		return out
	case []any:
		out := make([]any, len(v))
		for i, vv := range v {
			out[i] = canonical(vv, runIDMap)
		}
		return out
	case string:
		s := v
		for raw, repl := range runIDMap {
			if raw != "" && strings.Contains(s, raw) {
				s = strings.ReplaceAll(s, raw, repl)
			}
		}
		return s
	default:
		return value
	}
}

func stableJSON(v any) string {
	// Sort map keys for deterministic encoding.
	b, _ := stableEncode(v)
	return string(b)
}

func stableEncode(v any) ([]byte, error) {
	switch t := v.(type) {
	case map[string]any:
		keys := make([]string, 0, len(t))
		for k := range t {
			keys = append(keys, k)
		}
		sort.Strings(keys)
		var b strings.Builder
		b.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				b.WriteByte(',')
			}
			kj, _ := json.Marshal(k)
			b.Write(kj)
			b.WriteByte(':')
			ej, _ := stableEncode(t[k])
			b.Write(ej)
		}
		b.WriteByte('}')
		return []byte(b.String()), nil
	case []any:
		var b strings.Builder
		b.WriteByte('[')
		for i, vv := range t {
			if i > 0 {
				b.WriteByte(',')
			}
			ej, _ := stableEncode(vv)
			b.Write(ej)
		}
		b.WriteByte(']')
		return []byte(b.String()), nil
	default:
		return json.Marshal(v)
	}
}

// JournalLine — one canonical journal entry.
type JournalLine struct {
	Kind    string
	Payload string // stable-key-encoded JSON of canonical payload
}

// Snapshot — a run's journal, canonicalised. Position encodes seq.
type Snapshot struct {
	RunID string
	Lines []JournalLine
}

// Equals — two snapshots are equal iff every (kind, payload) matches in order.
func (s Snapshot) Equals(o Snapshot) bool {
	if len(s.Lines) != len(o.Lines) {
		return false
	}
	for i := range s.Lines {
		if s.Lines[i] != o.Lines[i] {
			return false
		}
	}
	return true
}

// DiffEntry — one per-position difference.
type DiffEntry struct {
	Index int
	Op    string // "==", "!=", "<" (only in B), ">" (only in A)
	A, B  *JournalLine
}

// Diff — per-position diff list, omitting equal entries.
func (s Snapshot) Diff(o Snapshot) []DiffEntry {
	var out []DiffEntry
	n := len(s.Lines)
	if len(o.Lines) > n {
		n = len(o.Lines)
	}
	for i := 0; i < n; i++ {
		var a, b *JournalLine
		if i < len(s.Lines) {
			a = &s.Lines[i]
		}
		if i < len(o.Lines) {
			b = &o.Lines[i]
		}
		if a == nil {
			out = append(out, DiffEntry{Index: i, Op: "<", B: b})
		} else if b == nil {
			out = append(out, DiffEntry{Index: i, Op: ">", A: a})
		} else if *a != *b {
			out = append(out, DiffEntry{Index: i, Op: "!=", A: a, B: b})
		}
	}
	return out
}

// CaptureSnapshotOpts — knobs for CaptureSnapshot.
type CaptureSnapshotOpts struct {
	DeadlineMs       int
	CanonicalRunID   string
}

// CaptureSnapshot — stream a run's journal via SubscribeRun, canonicalise,
// and stop at the first terminal `run` entry (or deadline). Mirrors
// `tape.chaos.snapshot.capture` (Python).
func CaptureSnapshot(ctx context.Context, client *tape.Client, runID string, opts CaptureSnapshotOpts) (Snapshot, error) {
	deadlineMs := opts.DeadlineMs
	if deadlineMs <= 0 {
		deadlineMs = 5000
	}
	canonRid := opts.CanonicalRunID
	if canonRid == "" {
		canonRid = "run-1"
	}
	runIDMap := map[string]string{runID: canonRid}

	subCtx, cancel := context.WithTimeout(ctx, time.Duration(deadlineMs)*time.Millisecond)
	defer cancel()

	stream, err := client.PB().SubscribeRun(subCtx, &tapepb.SubscribeRunRequest{RunId: runID, FromSeq: 0})
	if err != nil {
		return Snapshot{RunID: runID}, err
	}
	var lines []JournalLine
	for {
		entry, err := stream.Recv()
		if err != nil {
			break
		}
		var payload any
		if err := json.Unmarshal([]byte(entry.PayloadJson), &payload); err != nil {
			payload = map[string]any{"_raw": entry.PayloadJson}
		}
		canon := canonical(payload, runIDMap)
		lines = append(lines, JournalLine{Kind: entry.Kind, Payload: stableJSON(canon)})
		if entry.Kind == "run" {
			if m, ok := payload.(map[string]any); ok {
				status, _ := m["status"].(string)
				if _, term := terminalRunStatuses[strings.ToLower(status)]; term {
					break
				}
			}
		}
	}
	return Snapshot{RunID: runID, Lines: lines}, nil
}

// ── DeepSnapshot ─────────────────────────────────────────────────────────

// DeepSnapshot — full-projection walk; catches body-drift the journal
// summary in Snapshot misses.
type DeepSnapshot struct {
	RunID       string
	Decisions   []string
	Effects     []string
	Obligations []string
}

// Equals — pairwise equality on all three projections.
func (a DeepSnapshot) Equals(b DeepSnapshot) bool {
	eq := func(x, y []string) bool {
		if len(x) != len(y) {
			return false
		}
		for i := range x {
			if x[i] != y[i] {
				return false
			}
		}
		return true
	}
	return eq(a.Decisions, b.Decisions) && eq(a.Effects, b.Effects) && eq(a.Obligations, b.Obligations)
}

// CaptureDeepOpts — knobs for CaptureDeep.
type CaptureDeepOpts struct {
	CanonicalRunID string
	MaxDecisions   int
}

// CaptureDeep — walks decisions by index, effects via the journal +
// GetEffect, obligations via ListObligations. Higher cost than
// CaptureSnapshot; use when body-level drift matters.
func CaptureDeep(ctx context.Context, client *tape.Client, runID string, opts CaptureDeepOpts) (DeepSnapshot, error) {
	canonRid := opts.CanonicalRunID
	if canonRid == "" {
		canonRid = "run-1"
	}
	maxDec := opts.MaxDecisions
	if maxDec <= 0 {
		maxDec = 1000
	}
	runIDMap := map[string]string{runID: canonRid}
	canonField := func(d map[string]any) string {
		return stableJSON(canonical(d, runIDMap))
	}

	out := DeepSnapshot{RunID: runID}

	for i := int64(0); i < int64(maxDec); i++ {
		got, err := client.GetDecision(ctx, runID, i)
		if err != nil || !got.Found {
			break
		}
		d := got.Decision
		out.Decisions = append(out.Decisions, canonField(map[string]any{
			"decision_index": d.DecisionIndex, "model": d.Model,
			"request_json": d.RequestJson, "response_json": d.ResponseJson,
			"policy_version": d.PolicyVersion, "rationale": d.Rationale,
		}))
	}

	// Effects: walk journal once to collect keys, then GetEffect each.
	seen := map[string]struct{}{}
	subCtx, cancel := context.WithTimeout(ctx, 3*time.Second)
	defer cancel()
	stream, err := client.PB().SubscribeRun(subCtx, &tapepb.SubscribeRunRequest{RunId: runID, FromSeq: 0})
	if err == nil {
		for {
			entry, err := stream.Recv()
			if err != nil {
				break
			}
			if entry.Kind == "effect" {
				var p map[string]any
				if json.Unmarshal([]byte(entry.PayloadJson), &p) == nil {
					if k, _ := p["idempotency_key"].(string); k != "" {
						seen[k] = struct{}{}
					}
				}
			}
			if entry.Kind == "run" {
				var p map[string]any
				_ = json.Unmarshal([]byte(entry.PayloadJson), &p)
				status, _ := p["status"].(string)
				if _, term := terminalRunStatuses[strings.ToLower(status)]; term {
					break
				}
			}
		}
	}
	keys := make([]string, 0, len(seen))
	for k := range seen {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		got, err := client.GetEffect(ctx, runID, k)
		if err != nil || !got.Found {
			continue
		}
		e := got.Effect
		out.Effects = append(out.Effects, canonField(map[string]any{
			"tool_name": e.ToolName, "idempotency_key": e.IdempotencyKey,
			"status": e.Status, "request_json": e.RequestJson,
			"response_json": e.ResponseJson, "error_json": e.ErrorJson,
			"semantics": e.Semantics, "dispatch_mode": e.DispatchMode,
			"business_key": e.BusinessKey, "connector": e.Connector,
			"external_ref": e.ExternalRef, "decision_index": e.DecisionIndex,
		}))
	}

	// Obligations
	if resp, err := client.ListObligations(ctx, runID, false); err == nil {
		for _, o := range resp.Obligations {
			out.Obligations = append(out.Obligations, canonField(map[string]any{
				"kind": o.Kind, "effect_key": o.EffectKey, "status": o.Status,
				"payload_json": o.PayloadJson, "attempts": o.Attempts,
				"max_attempts": o.MaxAttempts, "last_error": o.LastError,
				"result_json": o.ResultJson, "compensator_ref": o.CompensatorRef,
			}))
		}
	}
	return out, nil
}

File: tape/sdk/typescript/src/chaos/snapshot.ts
// Snapshot — capture a run's journal + canonicalise for equality checks.
//
// Mirrors `tape.chaos.snapshot` from the Python SDK. The canonicalisation
// rules (strip timestamps, remap run-scoped identifiers) are kept in
// lockstep with the Python side so a Python-driven replay and a
// TS-driven replay are comparable.

import type { TapeClient } from '../client.ts';

const STRIP_KEYS: ReadonlySet<string> = new Set([
  'ts_ms', 'started_at_ms', 'ended_at_ms', 'last_update_time_ms',
  'lease_expires_at_ms', 'claim_expires_at_ms', 'dispatch_claim_expires_at_ms',
  'next_dispatch_at_ms', 'next_attempt_at_ms', 'fire_at_ms',
  'lease_owner', 'claimed_by', 'dispatch_claimed_by',
  'trace_id', 'span_id', 'parent_span_id',
  'seq', 'global_seq',
  // invocation_id varies per pass; remapping is sufficient for replay
  // semantics (matches the Python rule).
  'invocation_id',
]);

const TERMINAL_RUN_STATUSES: ReadonlySet<string> = new Set([
  'terminal', 'failed', 'cancelled', 'stuck',
]);

function canonical(value: unknown, runIdMap: Map<string, string>): unknown {
  if (value === null || value === undefined) return value;
  if (Array.isArray(value)) return value.map(v => canonical(v, runIdMap));
  if (typeof value === 'object') {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      if (STRIP_KEYS.has(k)) continue;
      out[k] = canonical(v, runIdMap);
    }
    return out;
  }
  if (typeof value === 'string') {
    let s = value;
    for (const [raw, replacement] of runIdMap) {
      if (raw && s.includes(raw)) s = s.split(raw).join(replacement);
    }
    return s;
  }
  return value;
}

function stableStringify(value: unknown): string {
  if (value === null || value === undefined) return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(',')}]`;
  if (typeof value === 'object') {
    const keys = Object.keys(value as object).sort();
    return `{${keys.map(k => `${JSON.stringify(k)}:${stableStringify((value as any)[k])}`).join(',')}}`;
  }
  return JSON.stringify(value);
}

export interface JournalLine {
  kind: string;
  payload: string;             // stable, sort-key-encoded JSON of canonical payload
}

export interface Snapshot {
  runId: string;
  lines: readonly JournalLine[];
  diff(other: Snapshot): Array<{ index: number; op: '==' | '!=' | '<' | '>'; a?: JournalLine; b?: JournalLine }>;
  equals(other: Snapshot): boolean;
}

function makeSnapshot(runId: string, lines: JournalLine[]): Snapshot {
  return {
    runId, lines: Object.freeze(lines),
    equals(other) {
      if (other.lines.length !== this.lines.length) return false;
      for (let i = 0; i < this.lines.length; i++) {
        if (this.lines[i].kind !== other.lines[i].kind) return false;
        if (this.lines[i].payload !== other.lines[i].payload) return false;
      }
      return true;
    },
    diff(other) {
      const out: Array<{ index: number; op: '==' | '!=' | '<' | '>'; a?: JournalLine; b?: JournalLine }> = [];
      const n = Math.max(this.lines.length, other.lines.length);
      for (let i = 0; i < n; i++) {
        const a = this.lines[i]; const b = other.lines[i];
        if (!a) out.push({ index: i, op: '<', b });
        else if (!b) out.push({ index: i, op: '>', a });
        else if (a.kind !== b.kind || a.payload !== b.payload) out.push({ index: i, op: '!=', a, b });
      }
      return out;
    },
  };
}

/**
 * Stream the journal for `runId` via SubscribeRun, canonicalise, and
 * stop at the first terminal `run` entry (or `deadlineMs`).
 */
export async function captureSnapshot(client: TapeClient, runId: string,
                                       opts: { deadlineMs?: number; canonicalRunId?: string } = {}): Promise<Snapshot> {
  const deadlineMs = opts.deadlineMs ?? 5_000;
  const runIdMap = new Map<string, string>([[runId, opts.canonicalRunId ?? 'run-1']]);
  const lines: JournalLine[] = [];
  const start = Date.now();

  const stream = client.subscribeRun({ runId, fromSeq: 0, timeoutMs: deadlineMs });
  try {
    for await (const entry of stream) {
      let payload: unknown;
      try { payload = JSON.parse((entry as any).payloadJson ?? '{}'); }
      catch { payload = { _raw: (entry as any).payloadJson }; }
      const canon = canonical(payload, runIdMap);
      lines.push({ kind: (entry as any).kind, payload: stableStringify(canon) });
      if ((entry as any).kind === 'run') {
        const status = String((payload as any)?.status ?? '').toLowerCase();
        if (TERMINAL_RUN_STATUSES.has(status)) break;
      }
      if (Date.now() - start > deadlineMs) break;
    }
  } finally {
    // The async iterator's finally already cancels the gRPC stream.
  }
  return makeSnapshot(runId, lines);
}

// ── DeepSnapshot ───────────────────────────────────────────────────────────

export interface DeepSnapshot {
  runId: string;
  decisions: readonly string[];   // canonicalised
  effects: readonly string[];
  obligations: readonly string[];
  equals(other: DeepSnapshot): boolean;
}

function canonField(d: Record<string, unknown>, runIdMap: Map<string, string>): string {
  return stableStringify(canonical(d, runIdMap));
}

/**
 * Walks the full projection tables — decisions by index, effects by
 * journal+GetEffect, obligations via ListObligations. Catches body-level
 * drift that the summary Snapshot misses.
 */
export async function captureDeep(client: TapeClient, runId: string,
                                   opts: { canonicalRunId?: string; maxDecisions?: number } = {}): Promise<DeepSnapshot> {
  const runIdMap = new Map<string, string>([[runId, opts.canonicalRunId ?? 'run-1']]);
  const maxDecisions = opts.maxDecisions ?? 1_000;

  // Decisions
  const decisions: string[] = [];
  for (let i = 0; i < maxDecisions; i++) {
    try {
      const got: any = await client.getDecision({ runId, decisionIndex: i });
      if (!got.found) break;
      const d = got.decision;
      decisions.push(canonField({
        decision_index: d.decisionIndex, model: d.model,
        request_json: d.requestJson, response_json: d.responseJson,
        policy_version: d.policyVersion, rationale: d.rationale,
      }, runIdMap));
    } catch { break; }
  }

  // Effects — walk journal once to collect keys, then GetEffect each.
  const seen = new Set<string>();
  const stream = client.subscribeRun({ runId, fromSeq: 0, timeoutMs: 3_000 });
  for await (const entry of stream) {
    if ((entry as any).kind === 'effect') {
      let p: any; try { p = JSON.parse((entry as any).payloadJson ?? '{}'); } catch { continue; }
      const k = String(p.idempotency_key ?? '');
      if (k) seen.add(k);
    } else if ((entry as any).kind === 'run') {
      let p: any; try { p = JSON.parse((entry as any).payloadJson ?? '{}'); } catch { p = {}; }
      if (TERMINAL_RUN_STATUSES.has(String(p.status ?? '').toLowerCase())) break;
    }
  }
  const effects: string[] = [];
  for (const key of [...seen].sort()) {
    try {
      const got: any = await client.getEffect({ runId, idempotencyKey: key });
      if (!got.found) continue;
      const e = got.effect;
      effects.push(canonField({
        tool_name: e.toolName, idempotency_key: e.idempotencyKey,
        status: e.status, request_json: e.requestJson,
        response_json: e.responseJson, error_json: e.errorJson,
        semantics: e.semantics, dispatch_mode: e.dispatchMode,
        business_key: e.businessKey, connector: e.connector,
        external_ref: e.externalRef, decision_index: e.decisionIndex,
      }, runIdMap));
    } catch { /* skip */ }
  }

  // Obligations
  const obligations: string[] = [];
  try {
    const resp: any = await client.listObligations({ runId, onlyUnresolved: false });
    for (const o of resp.obligations ?? []) {
      obligations.push(canonField({
        kind: o.kind, effect_key: o.effectKey, status: o.status,
        payload_json: o.payloadJson, attempts: o.attempts,
        max_attempts: o.maxAttempts, last_error: o.lastError,
        result_json: o.resultJson, compensator_ref: o.compensatorRef,
      }, runIdMap));
    }
  } catch { /* skip */ }

  const dec = Object.freeze(decisions);
  const eff = Object.freeze(effects);
  const obl = Object.freeze(obligations);
  const eqStr = (a: readonly string[], b: readonly string[]) =>
    a.length === b.length && a.every((x, i) => x === b[i]);
  return {
    runId,
    decisions: dec,
    effects: eff,
    obligations: obl,
    equals(other: DeepSnapshot) {
      return eqStr(dec, other.decisions)
          && eqStr(eff, other.effects)
          && eqStr(obl, other.obligations);
    },
  };
}
